### PR TITLE
[Feature] Data defined form widget editable state.

### PR DESCRIPTION
**Pull request by elpaso**
## Data-defined expression for widget editable status

The "Editable" checkbox status can now be controlled through an expression, the expression supports the "form" context, meaning that it can use `current_value( '<field_name>' )` to dynamically change the editable status according to changes to other fields in the form.

editable-data-defined

Funded by: Kanton Solothurn

### CODE REVIEW

File: src/gui/qgsattributeform.cpp [hunk 1]
```diff
@@ -1789,6 +1814,14 @@ void QgsAttributeForm::init()
                 mLabelDataDefinedProperties[ label ] = property;
               }
             }
+            if ( mLayer->editFormConfig().dataDefinedFieldProperties( fieldName ).hasProperty( QgsEditFormConfig::DataDefinedProperty::Editable ) )
+            {
+              const QgsProperty property { mLayer->editFormConfig().dataDefinedFieldProperties( fieldName ).property( QgsEditFormConfig::DataDefinedProperty::Editable ) };
+              if ( property.isActive() && ! property.expressionString().isEmpty() )
```
Comment (nyalldawson) [inline src/gui/qgsattributeform.cpp:1820]: If the property is a field based property, then the expressionString will always be empty and this logic won't work.

```suggestion
              if ( property.isActive() )
```

File: src/gui/qgsattributeform.cpp [hunk 2]
```diff
@@ -1885,6 +1918,15 @@ void QgsAttributeForm::init()
         formWidget->createSearchWidgetWrappers( mContext );
 
         label->setBuddy( eww->widget() );
+
+        if ( mLayer->editFormConfig().dataDefinedFieldProperties( fieldName ).hasProperty( QgsEditFormConfig::DataDefinedProperty::Editable ) )
+        {
+          const QgsProperty property { mLayer->editFormConfig().dataDefinedFieldProperties( fieldName ).property( QgsEditFormConfig::DataDefinedProperty::Editable ) };
+          if ( property.isActive() && ! property.expressionString().isEmpty() )
```
Comment (nyalldawson) [inline src/gui/qgsattributeform.cpp:1925]:
```suggestion
          if ( property.isActive() )
```

File: src/gui/qgsattributeform.cpp [hunk 3]
```diff
@@ -2402,6 +2444,14 @@ QgsAttributeForm::WidgetInfo QgsAttributeForm::createWidgetFromDef( const QgsAtt
                   mLabelDataDefinedProperties[ mypLabel ] = property;
                 }
               }
+              if ( mLayer->editFormConfig().dataDefinedFieldProperties( fieldName ).hasProperty( QgsEditFormConfig::DataDefinedProperty::Editable ) )
+              {
+                const QgsProperty property { mLayer->editFormConfig().dataDefinedFieldProperties( fieldName ).property( QgsEditFormConfig::DataDefinedProperty::Editable ) };
+                if ( property.isActive() && ! property.expressionString().isEmpty() )
```
Comment (nyalldawson) [inline src/gui/qgsattributeform.cpp:2450]:
```suggestion
                if ( property.isActive() )
```

**Review by nyalldawson — APPROVED** [review]

**Comment by elpaso** [comment]
unrelated test failure PyQgsSettingsTreeNode ..............................***Failed    1.64 sec

**Pull request merged**